cirq/google/api/v2/results_test.py

```diff
@@ -0,0 +1,224 @@
+                                                 dtype=bool),
+                             },
+                             repetitions=4),
+        ],
+    ]
+    proto = v2.results_to_proto(trial_results, measurements)
+    assert isinstance(proto, result_pb2.Result)
+    assert len(proto.sweep_results) == 2
+    deserialized = v2.results_from_proto(proto, measurements)
+    assert len(deserialized) == 2
+    for sweep_results, expected in zip(deserialized, trial_results):
+        assert len(sweep_results) == len(expected)
+        for trial_result, expected_trial_result in zip(sweep_results, expected):
+            assert trial_result.params == expected_trial_result.params
+            assert trial_result.repetitions == expected_trial_result.repetitions
+            np.testing.assert_array_equal(
+                trial_result.measurements['foo'],
+                expected_trial_result.measurements['foo'])
+
+
+def test_results_to_proto_sweep_repetitions():
+    measurements = [v2.MeasureInfo('foo', [q(0, 0)], slot=0)]
+    trial_results = [[
+        cirq.TrialResult(params=cirq.ParamResolver({'i': 0}),
+                         measurements={
+                             'foo': np.array([[0]], dtype=bool),
+                         },
+                         repetitions=1),
+        cirq.TrialResult(params=cirq.ParamResolver({'i': 1}),
+                         measurements={
+                             'foo': np.array([[0], [1]], dtype=bool),
+                         },
+                         repetitions=2),
+    ]]
+    with pytest.raises(ValueError, match='different numbers of repetitions'):
+        v2.results_to_proto(trial_results, measurements)
+
+
+def test_results_from_proto_qubit_ordering():
+    measurements = [v2.MeasureInfo('foo', [q(0, 0), q(0, 1), q(0, 2)], slot=0)]
+    proto = result_pb2.Result()
+    sr = proto.sweep_results.add()
+    sr.repetitions = 8
+    pr = sr.parameterized_results.add()
+    pr.params.assignments.update({'i': 0})
+    mr = pr.measurement_results.add()
+    mr.key = 'foo'
+    for qubit, results in [
+        (q(0, 1), 0b1100_1100),
+        (q(0, 2), 0b1010_1010),
+        (q(0, 0), 0b1111_0000),
+    ]:
+        qmr = mr.qubit_measurement_results.add()
+        qmr.qubit.id = qubit.proto_id()
+        qmr.results = bytes([results])
+
+    trial_results = v2.results_from_proto(proto, measurements)
+    trial = trial_results[0][0]
+    assert trial.params == cirq.ParamResolver({'i': 0})
+    assert trial.repetitions == 8
+    np.testing.assert_array_equal(
+        trial.measurements['foo'],
+        np.array([
+            [0, 0, 0],
+            [0, 0, 1],
+            [0, 1, 0],
+            [0, 1, 1],
+            [1, 0, 0],
+            [1, 0, 1],
+            [1, 1, 0],
+            [1, 1, 1],
+        ],
+                 dtype=bool))
+
+
+def test_results_from_proto_duplicate_qubit():
+    measurements = [v2.MeasureInfo('foo', [q(0, 0), q(0, 1), q(0, 2)], slot=0)]
```

Comment:
You never pass anything except 0 into the `row` argument, so you can't have tested that part of the serialization.

Reply:
v2 serializes qubits as string qubit ids, so there's not a separate `row` field anymore.

Reply:
I went ahead and changed the tests to use some qubits on different rows. It's more realistic in any case :-)